third_party/move/move-vm/runtime/src/storage/environment.rs

```diff
@@ -0,0 +1,266 @@
+    ) -> VMResult<PartiallyVerifiedModule> {
+        if !VERIFIED_MODULES_V2.contains(module_hash) {
+            // For regular execution, we cache already verified modules. Note
+            // that this even caches verification for the published modules.
+            // This should be ok because as long as the hash is the same, the
+            // deployed bytecode and any dependencies are the same, and so the
+            // cached verification result can be used.
+            move_bytecode_verifier::verify_module_with_config(
+                &self.vm_config().verifier_config,
+                compiled_module.as_ref(),
+            )?;
+            check_natives(compiled_module.as_ref())?;
+
+            if let Some(verifier) = &self.verifier_extension {
+                verifier.verify_module(compiled_module.as_ref())?;
+            }
+
+            VERIFIED_MODULES_V2.put(*module_hash);
+        }
+
+        Ok(PartiallyVerifiedModule(compiled_module, module_size))
+    }
+
+    /// Creates a fully verified module by running dependency verification
+    /// pass. The caller must provide verified module dependencies.
+    pub fn build_verified_module(
+        &self,
+        partially_verified_module: PartiallyVerifiedModule,
+        immediate_dependencies: &[Arc<Module>],
+    ) -> VMResult<Module> {
+        dependencies::verify_module(
+            partially_verified_module.0.as_ref(),
+            immediate_dependencies.iter().map(|m| m.module()),
+        )?;
+        let result = Module::new(
+            &self.natives,
+            partially_verified_module.1,
+            partially_verified_module.0,
+            self.struct_name_index_map(),
+        );
+
+        // Note: loader V1 implementation does not set locations for this error.
+        result.map_err(|e| e.finish(Location::Undefined))
+    }
+
+    /// Deserializes bytes into a compiled module. In addition, returns the size
+    /// of the module and its hash.
+    pub fn deserialize_into_compiled_module(
+        &self,
+        bytes: &Bytes,
+    ) -> VMResult<(CompiledModule, usize, [u8; 32])> {
+        let compiled_module =
+            CompiledModule::deserialize_with_config(bytes, &self.vm_config().deserializer_config)
+                .map_err(|err| {
+                let msg = format!("Deserialization error: {:?}", err);
+                PartialVMError::new(StatusCode::CODE_DESERIALIZATION_ERROR)
+                    .with_message(msg)
+                    .finish(Location::Undefined)
+            })?;
+
+        let mut sha3_256 = Sha3_256::new();
+        sha3_256.update(bytes);
+        let module_hash: [u8; 32] = sha3_256.finalize().into();
+
+        Ok((compiled_module, bytes.len(), module_hash))
+    }
+
+    /// Returns ann error is module's address and name do not match the expected values.
+    /// In general, we enforce this is the case at module publish time.
+    #[inline]
+    pub fn paranoid_check_module_address_and_name(
```

Comment:
is this a new check added?

Reply:
Old check, encapsulated as a function to share between V1 and V2.